From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Toffikk <slodkitofik@gmail.com>
Date: Sun, 17 Jan 2021 15:37:52 -0600
Subject: [PATCH] Sugarcane Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index ccfb8a8dcd51bf998c79298a22e38cebb2be6d39..ac3d5a8b0af00e32c26404e2bafab41f2c7d503a 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1730,6 +1730,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Sugarcane start - add config to timings report
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getSugarcaneConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Sugarcane end
+
         /**
          * Sends the component to the player
          *
