From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Toffikk <slodkitofik@gmail.com>
Date: Sun, 17 Jan 2021 15:37:52 -0600
Subject: [PATCH] Sugarcane Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 2f9ee4cf3b2163dbe086b4d40179bfa64ebfdecf..84d41cbaa149503fee69296c333b70dd8969b880 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1649,6 +1649,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Sugarcane start - add config to timings report
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getSugarcaneConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Sugarcane end
+
         /**
          * Sends the component to the player
          *
