From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Toffikk <slodkitofik@gmail.com>
Date: Sun, 17 Jan 2021 15:37:52 -0600
Subject: [PATCH] Sugarcane Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 80c70522eecb08c0dbfadcdeb21c9cfac42d7902..22723684664526bee6a586f04bc4aeaa627bf18e 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1702,6 +1702,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Sugarcane start - add config to timings report
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getSugarcaneConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Sugarcane end
+
         /**
          * Sends the component to the player
          *
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 42c9911885b1323032cc14c737c562695692744c..713d3e180b67a711359d3e5bb49f281ec4bf666b 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1570,7 +1570,7 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @throws IllegalArgumentException Thrown if the hash is not 20 bytes
      *     long.
      */
-    void setResourcePack(String url, byte[] hash, boolean force, net.kyori.adventure.text.Component prompt);
+    void setResourcePack(@NotNull String url, @NotNull byte[] hash, boolean force, net.kyori.adventure.text.Component prompt);
     // Purpur end
 
     /**
