From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Toffikk <slodkitofik@gmail.com>
Date: Sun, 17 Jan 2021 15:37:52 -0600
Subject: [PATCH] Sugarcane Config


diff --git a/build.gradle.kts b/build.gradle.kts
index 39dacd6b507c4afc997a7ed7b75ea83155f94708..396b8cc3f2288468d994584bb5bff7bdcc9e8e40 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -30,7 +30,9 @@ dependencies {
     api("com.google.code.gson:gson:2.8.8") // Paper - bump
     api("net.md-5:bungeecord-chat:1.16-R0.4")
     api("org.yaml:snakeyaml:1.28")
-    api("com.googlecode.json-simple:json-simple:1.1.1") // Paper
+    api("com.googlecode.json-simple:json-simple:1.1.1") { // Paper // Sugarcane
+        exclude(group = "junit", module = "junit")
+    }
     api("it.unimi.dsi:fastutil:8.2.2")
     apiAndDocs(platform("net.kyori:adventure-bom:$adventureVersion"))
     apiAndDocs("net.kyori:adventure-api")
@@ -71,12 +73,14 @@ configure<PublishingExtension> {
     }
 }
 
-val generateApiVersioningFile by tasks.registering {
-    val pomProps = layout.buildDirectory.file("pom.properties")
-    outputs.file(pomProps)
+val generateApiVersioningFile by tasks.registering { // Sugarcane
+    val pomProps = providers.fileContents(layout.buildDirectory.file("pom.properties"))
     doLast {
-        pomProps.get().asFile.writeText("version=${project.version}")
+        File("$pomProps").bufferedWriter().use { out -> 
+        out.write("version=${project.version}")
+       }
     }
+  outputs.cacheIf { true }
 }
 
 tasks.jar {
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index ccfb8a8dcd51bf998c79298a22e38cebb2be6d39..ac3d5a8b0af00e32c26404e2bafab41f2c7d503a 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1730,6 +1730,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Sugarcane start - add config to timings report
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getSugarcaneConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Sugarcane end
+
         /**
          * Sends the component to the player
          *
