From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Toffikk <slodkitofik@gmail.com>
Date: Sun, 17 Jan 2021 15:37:52 -0600
Subject: [PATCH] Sugarcane Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 80c70522eecb08c0dbfadcdeb21c9cfac42d7902..22723684664526bee6a586f04bc4aeaa627bf18e 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1702,6 +1702,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Sugarcane start - add config to timings report
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getSugarcaneConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Sugarcane end
+
         /**
          * Sends the component to the player
          *
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 743883f3d95b94a58ed7826dd74b3f37be9aecb9..2f80cdd9c21c9edaf8d354d2a8ce4352e9c814a4 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1530,7 +1530,7 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @throws IllegalArgumentException Thrown if the hash is not 20 bytes
      *     long.
      */
-    void setResourcePack(@NotNull String url, @NotNull byte[] hash, boolean force, String prompt);
+    void setResourcePack(@NotNull String url, @NotNull byte[] hash, boolean force, @NotNull String prompt);
 
     /**
      * Request that the player's client download and switch resource packs.
@@ -1570,7 +1570,7 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @throws IllegalArgumentException Thrown if the hash is not 20 bytes
      *     long.
      */
-    void setResourcePack(String url, byte[] hash, boolean force, net.kyori.adventure.text.Component prompt);
+    void setResourcePack(@NotNull String url, @NotNull byte[] hash, boolean force, @Nullable net.kyori.adventure.text.Component prompt);
     // Purpur end
 
     /**
