From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Toffikk <slodkitofik@gmail.com>
Date: Sun, 17 Jan 2021 15:37:52 -0600
Subject: [PATCH] Sugarcane Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index b8c380dae58d47514d684a6de2fd17b7d5135ab1..a244c3a8a91b2931acaa45b10bfb1e2e45c3448f 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1700,6 +1700,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Sugarcane start - add config to timings report
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getSugarcaneConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Sugarcane end
+
         /**
          * Sends the component to the player
          *
