From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Toffikk <slodkitofik@gmail.com>
Date: Tue, 7 Dec 2021 17:31:08 -0500
Subject: [PATCH] Sugarcane Config


diff --git a/build.gradle.kts b/build.gradle.kts
index 9203d7aa155319abfab0a65d4ef6804cb7e48769..e292faaaec5b0c751c6c2fad517910fcb980ecea 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -29,8 +29,8 @@ dependencies {
     api("net.md-5:bungeecord-chat:1.16-R0.4")
     api("org.yaml:snakeyaml:1.30")
     // Paper start
-    api("com.googlecode.json-simple:json-simple:1.1.1") {
-        isTransitive = false // includes junit
+    api("com.googlecode.json-simple:json-simple:1.1.1") { // Paper // Sugarcane
+        exclude(group = "junit", module = "junit")
     }
     api("it.unimi.dsi:fastutil:8.5.6")
     apiAndDocs(platform("net.kyori:adventure-bom:$adventureVersion"))
@@ -73,14 +73,16 @@ configure<PublishingExtension> {
     }
 }
 
-val generateApiVersioningFile by tasks.registering {
+val generateApiVersioningFile by tasks.registering { // Sugarcane
     inputs.property("version", project.version)
-    val pomProps = layout.buildDirectory.file("pom.properties")
+    val pomProps: Provider<RegularFile>
+    pomProps = layout.buildDirectory.file("pom.properties")
     outputs.file(pomProps)
     val projectVersion = project.version
     doLast {
         pomProps.get().asFile.writeText("version=$projectVersion")
     }
+    outputs.cacheIf { true }
 }
 
 tasks.jar {
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1a38b9e616c6096d8d64c8abaa7178bc5f84e8d4..7fe8caa055528d6c86eba6b28540d148af01c866 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1805,6 +1805,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Sugarcane start - add config to timings report
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getSugarcaneConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Sugarcane end
+
         /**
          * Sends the component to the player
          *
