From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Steven Tracey <steven@nevets.tech>
Date: Wed, 9 Nov 2022 20:18:15 -0500
Subject: [PATCH] Build fixes


diff --git a/build.gradle.kts b/build.gradle.kts
index 9f5694204091e23c4771657127a06f98e27ad8f1..9736af986ee28e479636efd22dbadd30db6285e5 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -1,6 +1,7 @@
 plugins {
     `java-library`
     `maven-publish`
+    id("com.github.johnrengelman.shadow")
 }
 
 java {
@@ -28,9 +29,12 @@ dependencies {
     api("net.md-5:bungeecord-chat:1.16-R0.4-deprecated+build.6") // Paper
     api("org.yaml:snakeyaml:1.33")
     // Paper start
+    compileOnly("com.mojang:authlib:2.3.31") // Sugarcane
+    api("de.tr7zw:item-nbt-api:2.10.0") // Sugarcane
     api("com.googlecode.json-simple:json-simple:1.1.1") {
-        isTransitive = false // includes junit
+        exclude(group = "junit", module = "junit") // Sugarcane
     }
+
     api("it.unimi.dsi:fastutil:8.5.6")
     apiAndDocs(platform("net.kyori:adventure-bom:$adventureVersion"))
     apiAndDocs("net.kyori:adventure-api")
@@ -81,6 +85,7 @@ val generateApiVersioningFile by tasks.registering {
     doLast {
         pomProps.get().asFile.writeText("version=$projectVersion")
     }
+    outputs.cacheIf { true }
 }
 
 tasks.jar {
@@ -94,6 +99,10 @@ tasks.jar {
     }
 }
 
+tasks.shadowJar {
+    relocate("de.tr7zw.changeme.nbtapi", "de.tr7zw.nbtapi")
+}
+
 tasks.withType<Javadoc> {
     val options = options as StandardJavadocDocletOptions
     options.overview = "src/main/javadoc/overview.html"
