From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Steven Tracey <steven@nevets.tech>
Date: Wed, 9 Nov 2022 20:42:17 -0500
Subject: [PATCH] Sugarcane-config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index e43fef0152468944d8a33036344a43e95fe58476..5a662426c3487743544b29e9ca7f2f26e8c3e935 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1969,6 +1969,12 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Sugarcane start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getSugarcaneConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+
         /**
          * Sends the component to the player
          *
