From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TheArcaneBrony <myrainbowdash949@gmail.com>
Date: Wed, 6 Apr 2022 22:36:17 +0200
Subject: [PATCH] Last tick time API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 9b9af189b3865055570ba533b97c92c105cc5a37..11785cf5f0a633e4e5d53470d58818f11bcaec32 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2397,6 +2397,16 @@ public final class Bukkit {
         return server.isStopping();
     }
 
+    // Sugarcane start
+    @Deprecated
+    public static long getLastTickMs() {
+        return server.getLastTickMs();
+    }
+    @NotNull public static java.time.Duration getLastTickTime() {
+        return server.getLastTickTime();
+    }
+    // Sugarcane end
+
     /**
      * Returns the {@link com.destroystokyo.paper.entity.ai.MobGoals} manager
      *
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1c8a9a4ea4da15457eb85907464a25a28833fbc6..69ff64f8e59336f9761a56951c1483938b968f74 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2101,6 +2101,27 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     boolean isStopping();
 
+    // Sugarcane start
+    /**
+     * Returns the time the last tick took in milliseconds.
+     *
+     * @return long time value
+     * @deprecated newer method with java's Duration and a possibility to convert it from millis to something else
+     */
+    @Deprecated
+    default long getLastTickMs() {
+        return getLastTickTime().toMillis();
+    }
+
+    /**
+     * Returns the time in {@link java.time.Duration} the last tick took.
+     *
+     * @return duration
+     */
+    @NotNull
+    java.time.Duration getLastTickTime();
+    // Sugarcane end
+
     /**
      * Returns the {@link com.destroystokyo.paper.entity.ai.MobGoals} manager
      *
