From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TheArcaneBrony <myrainbowdash949@gmail.com>
Date: Wed, 6 Apr 2022 22:36:17 +0200
Subject: [PATCH] Last tick time API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 6ba3272a10e357de99b75a7a7a63047f34a7fb8d..7b0230615ab24e99c628fa7f82b9108326d4fe81 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2290,6 +2290,16 @@ public final class Bukkit {
         return server.isStopping();
     }
 
+    // Sugarcane start
+    @Deprecated
+    public static long getLastTickMs() {
+        return server.getLastTickMs();
+    }
+    @NotNull public static java.time.Duration getLastTickTime() {
+        return server.getLastTickTime();
+    }
+    // Sugarcane end
+
     /**
      * Returns the {@link com.destroystokyo.paper.entity.ai.MobGoals} manager
      *
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 3fd5cdf1391c49567d923de882e96a63b7022e02..68bd90e51914a642b5864963ec11c19b3a8998cc 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2009,6 +2009,27 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     boolean isStopping();
 
+    // Sugarcane start
+    /**
+     * Returns the time the last tick took in milliseconds.
+     *
+     * @return long time value
+     * @deprecated newer method with java's Duration and a possibility to convert it from millis to something else
+     */
+    @Deprecated
+    default long getLastTickMs() {
+        return getLastTickTime().toMillis();
+    }
+
+    /**
+     * Returns the time in {@link java.time.Duration} the last tick took.
+     *
+     * @return duration
+     */
+    @NotNull
+    java.time.Duration getLastTickTime();
+    // Sugarcane end
+
     /**
      * Returns the {@link com.destroystokyo.paper.entity.ai.MobGoals} manager
      *
