From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Initial Source <auto@mated.null>
Date: Wed, 21 Jul 2021 22:22:41 +0200
Subject: [PATCH] Initial


diff --git a/build.gradle.kts b/build.gradle.kts
index 8f57361ecb47c531804563fc6ff9311e05b2a17f..887bbcdc4ed121b17bbb22f3df9de1db19351aee 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -10,6 +10,8 @@ import io.papermc.paperweight.util.openZip
 import io.papermc.paperweight.util.path
 import io.papermc.paperweight.util.registering
 import io.papermc.paperweight.util.set
+import io.papermc.paperweight.util.constants.DEOBF_NAMESPACE
+import io.papermc.paperweight.util.constants.SPIGOT_NAMESPACE
 import shadow.org.apache.logging.log4j.core.config.plugins.processor.PluginProcessor.PLUGIN_CACHE_FILE
 import java.nio.file.Files
 import java.text.SimpleDateFormat
@@ -122,8 +124,8 @@ val patchReobfMappings by tasks.registering<PatchMappings> {
     inputMappings.set(generateReobfMappings.flatMap { it.reobfMappings })
     patch.set(rootProject.layout.cache.resolve("paperweight/upstreams/paper/build-data/reobf-mappings-patch.tiny"))
 
-    fromNamespace.set(io.papermc.paperweight.util.Constants.DEOBF_NAMESPACE)
-    toNamespace.set(io.papermc.paperweight.util.Constants.SPIGOT_NAMESPACE)
+    fromNamespace.set(DEOBF_NAMESPACE)
+    toNamespace.set(SPIGOT_NAMESPACE)
 
     outputMappings.set(layout.cache.resolve("paperweight/mappings/reobf-patched.tiny"))
 }
