From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tr7zw <tr7zw@live.de>
Date: Thu, 2 Apr 2020 18:49:38 +0200
Subject: [PATCH] Add NBT API as a first-class lib

Original code by Yatopia, licensed under MIT License

diff --git a/build.gradle.kts b/build.gradle.kts
index 468411866430ffacf35c3e3225b6ff19217b4b94..53cf0b7931d634ac2038d39c44cba314afb84393 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -131,6 +131,7 @@ relocation {
     relocate("org.bukkit.craftbukkit" to "org.bukkit.craftbukkit.v$packageVersion") {
         exclude("org.bukkit.craftbukkit.Main*")
     }
+      relocate("de.tr7zw.changeme.nbtapi" to "de.tr7zw.nbtapi") // Sugarcane - Add NBT API as a first-class lib
 
     fun cb(pack: String) = "org.bukkit.craftbukkit.libs.$pack"
 
