From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tr7zw <tr7zw@live.de>
Date: Thu, 2 Apr 2020 18:49:38 +0200
Subject: [PATCH] Add NBT API as a first-class lib

Original code by Yatopia, licensed under MIT License

diff --git a/build.gradle.kts b/build.gradle.kts
index 1a21682209e32cc3939c27c65f60ff5feec00a57..2ae47f10e2126e3b2eb5079127326dd39ff3545d 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -155,6 +155,9 @@ tasks.shadowJar {
     relocate("org.bukkit.craftbukkit", "org.bukkit.craftbukkit.v$packageVersion") {
         exclude("org.bukkit.craftbukkit.Main*")
     }
+
+    relocate("de.tr7zw.changeme.nbtapi", "de.tr7zw.nbtapi")
+
     listOf(
         "jline", "org.apache.commons.codec", "org.apache.commons.io", // Paper - don't relocate fastutil
         "org.apache.commons.lang3", "org.apache.http", "org.apache.maven", "org.codehaus.plexus",
