From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tr7zw <tr7zw@live.de>
Date: Wed, 5 Aug 2020 14:25:50 -0500
Subject: [PATCH] Add GameProfileLookupEvent

Original code by Yatopia, licensed under MIT License

diff --git a/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java b/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java
index ef6aab69daa7ab952408b573bff6c6996cbc349a..c99f860c2f205c5487431ef95d5a4bec2aba7454 100644
--- a/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java
+++ b/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java
@@ -189,9 +189,17 @@ public class CraftPlayerProfile implements PlayerProfile {
 
         boolean isCompleteFromCache = this.completeFromCache(true, onlineMode);
         if (onlineMode && (!isCompleteFromCache || textures && !hasTextures())) {
-            GameProfile result = server.getSessionService().fillProfileProperties(profile, true);
-            if (result != null) {
-                copyProfileProperties(result, this.profile, true);
+            // Sugarcane start
+            org.sugarcanemc.sugarcane.events.GameProfileLookupEvent event = new org.sugarcanemc.sugarcane.events.GameProfileLookupEvent(!org.bukkit.Bukkit.isPrimaryThread(), profile.getId(), profile.getName());
+            org.bukkit.Bukkit.getPluginManager().callEvent(event);
+            GameProfile eventProfile = event.getGameProfile();
+            if (eventProfile != null) {
+                GameProfile result = eventProfile;
+            } else {
+                GameProfile result = server.getSessionService().fillProfileProperties(profile, true);
+                if (result != null) {
+                    copyProfileProperties(result, this.profile, true);
+                }
             }
             if (this.profile.isComplete()) {
                 server.getProfileCache().add(this.profile);
diff --git a/src/main/java/net/minecraft/world/level/block/entity/SkullBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SkullBlockEntity.java
index 96863338e1642105f71e681ec2b32c882b3ea342..948c58d20405ec7628dee6fdcd0a06f9757dc484 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SkullBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SkullBlockEntity.java
@@ -156,7 +156,15 @@ public class SkullBlockEntity extends BlockEntity {
                 Property property = (Property) Iterables.getFirst(gameprofile1.getProperties().get("textures"), (Object) null);
 
                 if (property == null) {
-                    gameprofile1 = SkullBlockEntity.sessionService.fillProfileProperties(gameprofile1, true);
+                    // Sugarcane start
+                    org.sugarcanemc.sugarcane.events.GameProfileLookupEvent event = new org.sugarcanemc.sugarcane.events.GameProfileLookupEvent(!org.bukkit.Bukkit.isPrimaryThread(), gameprofile1.getId(), gameprofile1.getName());
+                    org.bukkit.Bukkit.getPluginManager().callEvent(event);
+                    GameProfile eventProfile = event.getGameProfile();
+                    if (eventProfile != null) {
+                        gameprofile1 = eventProfile;
+                    } else {
+                        gameprofile1 = SkullBlockEntity.sessionService.fillProfileProperties(gameprofile1, true);
+                    }
                 }
 
                 GameProfile finalGameprofile = gameprofile1;
