From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TheArcaneBrony <myrainbowdash949@gmail.com>
Date: Wed, 6 Apr 2022 23:02:15 +0200
Subject: [PATCH] TileEntity HashedList


diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index 64cbb68924f80f7d44aba99f1d515c347515a68c..8647a62fd902372d2033c62217c4bd744c1178cf 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -115,8 +115,9 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
     public static final int TICKS_PER_DAY = 24000;
     public static final int MAX_ENTITY_SPAWN_Y = 20000000;
     public static final int MIN_ENTITY_SPAWN_Y = -20000000;
-    protected final List<TickingBlockEntity> blockEntityTickers = Lists.newArrayList(); public final int getTotalTileEntityTickers() { return this.blockEntityTickers.size(); } // Paper
-    private final List<TickingBlockEntity> pendingBlockEntityTickers = Lists.newArrayList();
+    public final List<TickingBlockEntity> blockEntityTickers = org.sugarcanemc.sugarcane.lithiumclasses.HashedList.wrapper(Lists.newArrayList());
+    public final int getTotalTileEntityTickers() { return this.blockEntityTickers.size(); } // Paper // Sugarcane
+    private final List<TickingBlockEntity> pendingBlockEntityTickers = org.sugarcanemc.sugarcane.lithiumclasses.HashedList.wrapper(Lists.newArrayList()); // Sugarcane
     private boolean tickingBlockEntities;
     public final Thread thread;
     private final boolean isDebug;
