From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Wed, 20 Jan 2021 16:36:48 -0500
Subject: [PATCH] add config for logging login location

Original code by Yatopia, licensed under MIT License

diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 2f529c25ffcd6c9d0e9f8ffbea3dbb9f76ad8017..43235fd01bd52c7de8a17076b964ea1ad778b700 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -482,7 +482,14 @@ public abstract class PlayerList {
         }
         // Paper end
         // CraftBukkit - Moved from above, added world
-        PlayerList.LOGGER.info("{}[{}] logged in with entity id {} at ([{}]{}, {}, {})", player.getName().getString(), s1, player.getId(), worldserver1.serverLevelData.getLevelName(), player.getX(), player.getY(), player.getZ());
+
+        // Sugarcane start - configurable logging of player login location
+        if (org.sugarcanemc.sugarcane.config.SugarcaneConfig.logPlayerLoginLoc) {
+            PlayerList.LOGGER.info("{}[{}] logged in with entity id {} at ([{}]{}, {}, {})", player.getName().getString(), s1, player.getId(), worldserver1.serverLevelData.getLevelName(), player.getX(), player.getY(), player.getZ());
+        } else {
+            PlayerList.LOGGER.info("{}[{}] logged in with entity id {}", player.getName().getString(), s1, player.getId());
+        }
+        // Sugarcane end - configurable logging of player login location
     }
 
     public void updateEntireScoreboard(ServerScoreboard scoreboard, ServerPlayer player) {
diff --git a/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java b/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java
index 197a5f5aaf2adfab79005d1acc6d26039f7239bb..dc2f698ea72cb32bf4714d95b3e31f8b76c2dae0 100644
--- a/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java
+++ b/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java
@@ -170,6 +170,10 @@ public class SugarcaneConfig {
 		return config.getString(path, config.getString(path));
 }
 
+public static boolean logPlayerLoginLoc = true;
+	private static void general() {
+		logPlayerLoginLoc = getBoolean("settings.log-player-login-location", logPlayerLoginLoc);
+	}
 	public static boolean fixProtocolLib = true;
 	private static void protocolLib() {
 		fixProtocolLib = getBoolean("settings.fix-protocollib", fixProtocolLib);
