From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Wed, 20 Jan 2021 16:36:48 -0500
Subject: [PATCH] add config for logging login location

Original code by Yatopia, licensed under MIT License

diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 420a0011ef222169ea5b356d761164bd4debcc5a..7182f45f368b8e31345b3824876a00da2f48be81 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -484,7 +484,14 @@ public abstract class PlayerList {
         }
         // Paper end
         // CraftBukkit - Moved from above, added world
-        PlayerList.LOGGER.info("{}[{}] logged in with entity id {} at ([{}]{}, {}, {})", player.getName().getString(), s1, player.getId(), worldserver1.serverLevelData.getLevelName(), player.getX(), player.getY(), player.getZ());
+
+        // Sugarcane start - configurable logging of player login location
+        if (org.sugarcanemc.sugarcane.config.SugarcaneConfig.logPlayerLoginLoc) {
+            PlayerList.LOGGER.info("{}[{}] logged in with entity id {} at ([{}]{}, {}, {})", player.getName().getString(), s1, player.getId(), worldserver1.serverLevelData.getLevelName(), player.getX(), player.getY(), player.getZ());
+        } else {
+            PlayerList.LOGGER.info("{}[{}] logged in with entity id {}", player.getName().getString(), s1, player.getId());
+        }
+        // Sugarcane end - configurable logging of player login location
     }
 
     public void updateEntireScoreboard(ServerScoreboard scoreboard, ServerPlayer player) {
diff --git a/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java b/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java
index 87f06b075f7b33032c1acda3a4157c5cb6ec7818..c10f930391b2cd16cd91d175479e09ed68b6c053 100644
--- a/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java
+++ b/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java
@@ -171,6 +171,10 @@ public class SugarcaneConfig {
 		return config.getString(path, config.getString(path));
 }
 
+public static boolean logPlayerLoginLoc = true;
+	private static void general() {
+		logPlayerLoginLoc = getBoolean("settings.log-player-login-location", logPlayerLoginLoc);
+	}
 	public static boolean fixProtocolLib = true;
 	private static void protocolLib() {
 		fixProtocolLib = getBoolean("settings.fix-protocollib", fixProtocolLib);
