From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tofik=20=E2=99=A1?= <slodkitofiK@gmail.com>
Date: Wed, 23 Jun 2021 14:43:59 +0200
Subject: [PATCH] Move configs to config dir


diff --git a/src/main/java/gg/airplane/AirplaneConfig.java b/src/main/java/gg/airplane/AirplaneConfig.java
index 6fc23a85ddd66a69b83684cdb602b6a312e9d72b..8ae8ace23f6d6b1ca9196c461f83a6074d6089be 100644
--- a/src/main/java/gg/airplane/AirplaneConfig.java
+++ b/src/main/java/gg/airplane/AirplaneConfig.java
@@ -16,7 +16,7 @@ public class AirplaneConfig {
     private static AIR config;
 
     public static void load() throws IOException {
-        File configFile = new File("airplane.air");
+        File configFile = new File("config/airplane.air");
         if (configFile.exists()) {
             try (FileInputStream inputStream = new FileInputStream(configFile)) {
                 config = new AIR(inputStream);
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index e955fabb657a5e199e2efc8ed52b143f4060f564..e3594ee28f2de85196f46968ac3d7834ef23f791 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -110,13 +110,13 @@ public class Main {
                 acceptsAll(Main.asList("b", "bukkit-settings"), "File for bukkit settings")
                         .withRequiredArg()
                         .ofType(File.class)
-                        .defaultsTo(new File("bukkit.yml"))
+                        .defaultsTo(new File("config/bukkit.yml"))
                         .describedAs("Yml file");
 
                 acceptsAll(Main.asList("C", "commands-settings"), "File for command settings")
                         .withRequiredArg()
                         .ofType(File.class)
-                        .defaultsTo(new File("commands.yml"))
+                        .defaultsTo(new File("config/commands/commands.yml"))
                         .describedAs("Yml file");
 
                 acceptsAll(Main.asList("forceUpgrade"), "Whether to force a world upgrade");
@@ -135,7 +135,7 @@ public class Main {
                 acceptsAll(Main.asList("S", "spigot-settings"), "File for spigot settings")
                         .withRequiredArg()
                         .ofType(File.class)
-                        .defaultsTo(new File("spigot.yml"))
+                        .defaultsTo(new File("config/spigot.yml"))
                         .describedAs("Yml file");
                 // Spigot End
 
@@ -143,14 +143,14 @@ public class Main {
                 acceptsAll(asList("paper", "paper-settings"), "File for paper settings")
                         .withRequiredArg()
                         .ofType(File.class)
-                        .defaultsTo(new File("paper.yml"))
+                        .defaultsTo(new File("config/paper.yml"))
                         .describedAs("Yml file");
                 // Paper end
                 // Tuinity start - Server Config
                 acceptsAll(asList("tuinity", "tuinity-settings"), "File for tuinity settings")
                         .withRequiredArg()
                         .ofType(File.class)
-                        .defaultsTo(new File("tuinity.yml"))
+                        .defaultsTo(new File("config/tuinity.yml"))
                         .describedAs("Yml file");
                 // Tuinity end - Server Config
 
@@ -158,7 +158,7 @@ public class Main {
                 acceptsAll(asList("Sugarcane", "Sugarcane-settings"), "File for Sugarcane settings")
                         .withRequiredArg()
                         .ofType(File.class)
-                        .defaultsTo(new File("sugarcane.yml"))
+                        .defaultsTo(new File("config/sugarcane.yml"))
                         .describedAs("Yml file");
                 // Sugarcane end
 
