From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Steven Tracey <steven@nevets.tech>
Date: Tue, 7 Dec 2021 21:12:11 -0500
Subject: [PATCH] Item stuck sleep config


diff --git a/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java b/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java
index 99e8eb7eaa784ecfd89ffb3dc04c8af761a24bc3..5735813d72ae98aa6f72fc9e1813f08c9408cc63 100644
--- a/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java
+++ b/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java
@@ -197,4 +197,9 @@ private static void brandName() {
 		checkFlying = getBoolean("settings.checks.flight", checkFlying);
 		checkVehicleFlying = getBoolean("settings.checks.vehicle-flight", checkVehicleFlying);
 	}
+
+	public static int itemStuckSleepTicks = 1;
+	private static void itemStuckSleepTicks() {
+		itemStuckSleepTicks = getInt("settings.itemStuckSleepTicks", 1);
+	}
 }
\ No newline at end of file
