From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: greysilly7 <greysilly7@gmail.com>
Date: Sun, 27 Jun 2021 17:18:36 -0400
Subject: [PATCH] Awooooo


diff --git a/src/main/java/net/minecraft/server/Eula.java b/src/main/java/net/minecraft/server/Eula.java
index e76973410c663234cef5e6709ff5b5bd24c5cddc..6afc36e5924d1bb90028ca7ac5eb7a9f3be7d83d 100644
--- a/src/main/java/net/minecraft/server/Eula.java
+++ b/src/main/java/net/minecraft/server/Eula.java
@@ -78,7 +78,7 @@ public class Eula {
                 try {
                     Properties properties = new Properties();
                     properties.setProperty("eula", "false");
-                    properties.store(outputStream, "By changing the setting below to TRUE you are indicating your agreement to our EULA (https://account.mojang.com/documents/minecraft_eula).");  // Paper - fix lag; // Tuinity - Tacos are disgusting
+                    properties.store(outputStream, "By changing the setting below to TRUE you are indicating your agreement to our EULA (https://account.mojang.com/documents/minecraft_eula). Awooo"); // Sugarcane // Paper - fix lag; // Tuinity - Tacos are disgusting
                 } catch (Throwable var5) {
                     if (outputStream != null) {
                         try {
