From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tr7zw <tr7zw@live.de>
Date: Fri, 31 Jul 2020 21:48:14 -0500
Subject: [PATCH] Item stuck sleep config

Original code by Yatopia, licensed under MIT License

diff --git a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
index 9ec17c9b469dd058cf382164027f0392a92b2c29..6c4e661cfb4c556c07222d1ac4d3ccc40157c108 100644
--- a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
@@ -132,7 +132,7 @@ public class ItemEntity extends Entity {
 
             if (this.level.isClientSide) {
                 this.noPhysics = false;
-            } else {
+            } else if (!this.onGround || this.noPhysics || this.tickCount % org.sugarcanemc.sugarcane.config.SugarcaneConfig.itemStuckSleepTicks == 0) { // Sugarcane - Item stuck sleep config
                 this.noPhysics = !this.level.noCollision((Entity) this, this.getBoundingBox().deflate(1.0E-7D), (entity) -> {
                     return true;
                 });
diff --git a/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java b/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java
index 59aeb2545bf283974b7cc4333f00025e3d84950f..2749fdf41742effee615ff78b78fa339e8366331 100644
--- a/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java
+++ b/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java
@@ -196,4 +196,9 @@ private static void brandName() {
         checkFlying = getBoolean("settings.checks.flight", checkFlying);
         checkVehicleFlying = getBoolean("settings.checks.vehicle-flight", checkVehicleFlying);
     }
+
+    public static int itemStuckSleepTicks = 1;
+    private static void itemStuckSleepTicks() {
+        itemStuckSleepTicks = getInt("settings.itemStuckSleepTicks", 1);
+    }
 }
\ No newline at end of file
