From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TheArcaneBrony <myrainbowdash949@gmail.com>
Date: Mon, 7 Feb 2022 04:14:50 +0000
Subject: [PATCH] Ignore feature being null in world gen


diff --git a/src/main/java/net/minecraft/world/level/chunk/ChunkGenerator.java b/src/main/java/net/minecraft/world/level/chunk/ChunkGenerator.java
index 096f05db09d98e2b8edfd35d2cd97f9b4085d056..564fa9ff36d60c4feb5774a1d5f686afaa391702 100644
--- a/src/main/java/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/src/main/java/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -458,6 +458,8 @@ public abstract class ChunkGenerator implements BiomeManager.NoiseBiomeSource {
             if (structurestart == null || !structurestart.isValid()) {
                 StructureStart<?> structurestart1 = StructureFeatures.STRONGHOLD.generate(registryManager, this, this.biomeSource, structureManager, worldSeed, chunkcoordintpair, ChunkGenerator.fetchReferences(structureAccessor, chunk, sectionposition, StructureFeature.STRONGHOLD), structuresettingsfeature, chunk, ChunkGenerator::validStrongholdBiome);
 
+                // Sugarcane - mute features being null
+                if(structurestart1.getFeature() != null)
                 // Purpur start
                 if (new org.purpurmc.purpur.event.world.StructureGenerateEvent(
                         structureAccessor.getWorld().getWorld(),
